### ISSUE TITLE
The new deactivation process (due sanity checks) is preventing parts with TweakScale deactivated to be attached

### ISSUE THREAD

**Issue by Lisias** [net-lisias-ksp/TweakScale#125]
Oh, well...

You can't attach parts with a deactivated TweakScale module. Apparently, I did a too good of a job on deactivating the thing. (sigh).

I wondering  what changed since the last time I tested this (a few weeks ago), I remember attaching a tube (one of the deactivated parts) on something to test if it was being scaled or not... :/

Debug time.

**Comment by Lisias** [comment]
Well, this is the (long) story.

Some parts just cannot be scaled (for many reasons, some will be fixed in the future, others are beyound my reach). So some parts will have TweakScale deactivated, and life goes on.

Until 1.7.3 , bluntly removing the Module from the `prefab.Modules` was the easiest and safest way - you would plain "unpatch" the part, removing any TweakScale traces from it. Clear, simple, easy.

But on 1.8 something changed (probably the modules are now being Started on the Main Menu, while they were being Started on the start of a new or loaded game). And so things started to bork terribly.

So I coded a brute force deactivation on commit https://github.com/net-lisias-ksp/TweakScale/commit/2465deb2d2ae9415bdf187292ef7dda27b68a459 for issue #114 . This did the trick, and it worked at that time for sure (otherwise I could not had tested the solution - I need to attach a part in order to open the PAW and see if the Scale widget was there or not.

BUT... On KSP 1.10 the stunt broke again.

I managed to nail down the problem to the remove of the Module itself. Apparently (and I'm guessing) some code on KSP now reads the prefab's list of Modules **only once**, and before I can do my magic. And so, probably (and again) on the OnClone handling, borks because it used that old, out of date copy of that prefab and ends up crashing because that entry was removed and garbage collected already.

The solution is simple, besides far from ideal - just leave the Module there, wasting memory and CPU cycles, on a dry run. In order to avoid taxing older KSP where the stunt worked fine (and, so, avoiding wasting that resources on an unoptimised environment).

**Comment by Lisias** [comment]
I was wrong. I probably was blinded by Survivorship Bias, I think.

The current stunt didn't worked on all KSP versions. I have some theories, but right now I will extinguish the fire under my nose, and will try to find out why it started later.

**Comment by 7ranceaddic7** [comment]
I think it's time to make this suggestion. You need to start giving SERIOUS reconsideration to this concept: One TweakScale to rule them all.

It's seems there have been significant shifts in  process handling between v1.8, v1.9, and v1.10.  Each has introduced a significant quirk which compounds the difficulty of integrating a solution for the latest iteration while maintaining functionality of previous versions ALL WHILE IN A SINGULAR RELEASE.

I think it's becoming apparent that the more effective and efficient solution is to split TS into separate builds for each version. Any headaches are then isolated to their respective branch . If appropriate, fixes can always be merged across trees.

**Comment by Lisias** [comment]
> I think it's time to make this suggestion. You need to start giving SERIOUS reconsideration to this concept: One TweakScale to rule them all.

Perhaps, but not on this issue. Every single KSP version I tested with the current codetree borked the same, downto 1.4.1. Apparently, I was exploiting a singular behaviour that happened to work with the previous handling of events on TweakScale before I had to refactor it to adapt to Module Part Variants.

Having a dedicates code-tree for each version would not prevented the breakage due the Editor stunt on KSP 1.9 neither, and the problem being fixed on 1.10 corroborates this thesis.

Rewrite a new version of TweakScale for every new minor release of KSP is absurd, mainly because the only thing that the new KSP versions are really changing is the Editor (mis)behaviour and the UI that it's more convenient on KSP 1.8 and later. And none of that would had prevented this problem.

You misdiagnosed the problem. This time, **it is** a bug on TweakScale and the regression tests failing on every KSP from 1.10 down to 1.4.1 says me clearly that removing TweakScale from the prefab was working by "accident", and not due something related to KSP itself.

The real solution is to fix TweakScale - and the testing made clear that the fix is demanded for every KSP version since 1.4.1. Having multiple code-trees would just force me to apply the fix on multiple code-trees.

**Comment by 7ranceaddic7** [comment]
Yep.  What a difference 6h makes.

**Comment by Lisias** [comment]
> Yep. What a difference 6h makes.

And regression tests. :)

I have a test bed for every single KSP released since 1.4.0, and additional ones for 1.3.1, 1.2.2 and 1.1.3 .

And I test every bug on most of them in order to realise where it borks, where it works and so I try to figure out the cause.

Something that breaks only on some KSP versions I tag as an idiosyncrasy that must be handled on a KSP by KSP basis (and this will, someday, force splitting TweakScale on different code-trees - or creating another stunt on KSP-Recall).

Something breaking consistently on every KSP version is my fault. Point. :)

**Comment by Lisias** [comment]
Fixed on commit https://github.com/net-lisias-ksp/TweakScale/commit/5e1172e18e68985836bc362c873cbe673651459a

**Comment by Lisias** [comment]
> BUT... On KSP 1.10 the stunt broke again. _[edit **NO!** KSP appears to be innocent here!]_

No. I was wrong on thinking I was wrong. Well, I'm not alone (right, Einstein? :) )

That what's happening:

On KSP <= 1.7.3, removing the module from the prefab's part module list was, effectively, removing the module from that specific part's life cycle. This is what I was doing: 

```
                        Log.warn("Removing TweakScale support for {0} ({1}).", p.name, p.title);
                        prefab.Modules.Remove(prefab.Modules["TweakScale"]);
                        Log.error("Part {0} ({1}) didn't passed the sanity check due {2}.", p.name, p.title, r);
```

However, the module was being allowed to be kept alive on memory, doing nothing - the List atribute was exposed directly to the userland, something I'm biased to do not notice because when **I** write these kind of thing, I give a facade to the userland simulating a List, and the inner code of the facade would handle anything is needed to carry on the task.

Oukey, this part of the problem was diagnosed. This **was** working acceptably (besides not optimally) on KSP <= 1.7.3. So, yeah, something changed on KSP 1.8.0 on this matter (spoiler: changed for the best this time).

So I shoved this stunt into the codebase and rushed an emergencial release to have things kinda under control while I assess the situation (as I didn't had the information above at that time on hands):

```
                        Log.warn("Removing TweakScale support for {0} ({1}).", p.name, p.title);
                        (prefab.Modules["TweakScale"] as TweakScale).isActive = false; // Needed on KSP >= 1.8
                        prefab.Modules.Remove(prefab.Modules["TweakScale"]);
                        Log.error("Part {0} ({1}) didn't passed the sanity check due {2}.", p.name, p.title, r);
```

The `isActive` [stunt was a code rushed into production](https://github.com/net-lisias-ksp/TweakScale/commit/2465deb2d2ae9415bdf187292ef7dda27b68a459) meant to be used by the TweakScale's life cycle handlers to prevent it to be run, as well to prevent it to save the badly initialised and/or damaged module data to infect the craft file or savegame. It was kinda an overreaction, as setting the *canon* `enabled` would accomplish the same - but I choose to be redundant in order to help myself diagnosing things if something goes wrong on the wild.

However, this had the side effect of shoving up "RogueDuplicate" entries into the user's craft and savegames. (sigh). Annoying, but it was what kept everything sane - better safe and annoyed than annoyed by being sorry.

And, finally, this is what solved the problem on KSP >= 1.8.0:

```
                        Log.error("Part {0} ({1}) didn't passed the sanity check due {2}.", p.name, p.title, r);
                        Log.warn("Removing TweakScale support for {0} ({1}).", p.name, p.title);
                        prefab.RemoveModule(prefab.Modules ["TweakScale"]);
```

The `RemoveModule` code not only removes the module from the list, but does something else preventing it to accidentally being activated later : it **kills** the module :P - but now things start to get interesting, apparently this wasn't being done on KSP <= 1.7.3 because I got the same misbehaviours on it even after running this new code.

In the after math, this is the log of the DEBUG binary for TweakScale Beta (**warning**: this thing floods KSP.log with a myriad of usually useless data, impairing the performance) on KSP <= 1.7.3:

```
[LOG 02:16:00.135] [TweakScale] TRACE: Found part named FSfloatEndTail ; title FS4SFT Seaplane Float (Tail):
[LOG 02:16:00.135] [TweakScale] TRACE:  Part FSfloatEndTail has module FSrudder
[LOG 02:16:00.135] [TweakScale] TRACE:  Part FSfloatEndTail has module FSbuoyancy
[LOG 02:16:00.135] [TweakScale] TRACE:  Part FSfloatEndTail has module TweakScale
[LOG 02:16:00.135] [TweakScale] TRACE: Checking Sanity for FSfloatEndTail at Firespitter/Parts/Structural/FS_floatEndTail/part/FSfloatEndTail
[LOG 02:16:00.135] [TweakScale] ERROR: Part FSfloatEndTail (FS4SFT Seaplane Float (Tail)) didn't passed the sanity check due using FSbuoyancy module without TweakScaleComp
anion for Firespitter installed - see issue [#1] from TSC_FS ( https://github.com/net-lisias-ksp/TweakScaleCompantion_FS/issues/1 ). at error:0
[LOG 02:16:00.135] [TweakScale] WARNING: Removing TweakScale support for FSfloatEndTail (FS4SFT Seaplane Float (Tail)).
[LOG 02:16:00.135] [TweakScale] TRACE: Found part named FSfloatGearbay ; title FS4SFG Seaplane Float (Gear bay):
<blablabalablabla>
```

And this is the same code acting on KSP >= 1.8.0 :
```
[LOG 02:34:08.696] [TweakScale] TRACE: Found part named EnginePlate1p5 ; title EP-18 Engine Plate:
[LOG 02:34:08.696] [TweakScale] TRACE:  Part EnginePlate1p5 has module ModuleDynamicNodes
[LOG 02:34:08.696] [TweakScale] TRACE:  Part EnginePlate1p5 has module ModuleDecouple
[LOG 02:34:08.696] [TweakScale] TRACE:  Part EnginePlate1p5 has module ModulePartVariants
[LOG 02:34:08.696] [TweakScale] TRACE:  Part EnginePlate1p5 has module ModuleJettison
[LOG 02:34:08.696] [TweakScale] TRACE:  Part EnginePlate1p5 has module TweakScale
[LOG 02:34:08.696] [TweakScale] TRACE: Checking Sanity for EnginePlate1p5 at SquadExpansion/MakingHistory/Parts/Coupling/EnginePlate_1p5/EnginePlate1p5
[LOG 02:34:08.697] [TweakScale] ERROR: Part EnginePlate1p5 (EP-18 Engine Plate) didn't passed the sanity check due having a ModulePartVariants with Mass - see issue [#13](
 https://github.com/net-lisias-ksp/TweakScale/issues/13 ). at error:0
[LOG 02:34:08.698] [TweakScale] WARNING: Removing TweakScale support for EnginePlate1p5 (EP-18 Engine Plate).
[LOG 02:34:08.698] [TweakScale] TRACE: OnDestroy EnginePlate1p5
[LOG 02:34:08.698] [TweakScale] TRACE: Found part named EnginePlate2 ; title EP-25 Engine Plate:
```

See the `OnDestroy` log entry. It's not happening on KSP <= 1.7.3 using the same code :P

Right now, I'm toying with the **brute force destruction** of the module, getting rid of it **for good** as it was never patched by Module Manager or hardcoded on the part. It's currently only on TweakScale Beta, because I want to check against any collateral effects before shoving this stunt on a Release. I expect this will solve the problem for **EVERYBODY** and not only for newer KSP versions.

This is the same log, but after I cooked the destroy thingy on the KSP <= 1.7.3 :

```
[LOG 03:22:56.643] [TweakScale] WARNING: Removing TweakScale support for FSfloatEnd (FS4SFN Seaplane Float (Nose)).
[LOG 03:22:56.643] [TweakScale] TRACE: OnDestroy FSfloatEnd
[LOG 03:22:56.643] [TweakScale] TRACE: Found part named FSfloatEndTail ; title FS4SFT Seaplane Float (Tail):
[LOG 03:22:56.644] [TweakScale] TRACE:  Part FSfloatEndTail has module FSrudder
[LOG 03:22:56.644] [TweakScale] TRACE:  Part FSfloatEndTail has module FSbuoyancy
[LOG 03:22:56.644] [TweakScale] TRACE:  Part FSfloatEndTail has module TweakScale
[LOG 03:22:56.644] [TweakScale] TRACE: Checking Sanity for FSfloatEndTail at Firespitter/Parts/Structural/FS_floatEndTail/part/FSfloatEndTail
<blablabala>
```

And this is the code that made the trick:

```
                        Log.error("Part {0} ({1}) didn't passed the sanity check due {2}.", p.name, p.title, r);
                        Log.warn("Removing TweakScale support for {0} ({1}).", p.name, p.title);

                        PartModule m = prefab.Modules["TweakScale"];
                        prefab.RemoveModule(m);
                        if (KSPe.Util.KSP.Version.Current < KSPe.Util.KSP.Version.FindByVersion(1,8,0))
                            UnityEngine.Object.Destroy(m); 
```
 
Problem solved. Hopefully. :P
